Add conversion for aten.transpose

mcw-zwakeelTT opened the pull request:
### Problem description
Transformation for torch transpose op to ttnn equivalent.

### What's changed
Added conversion routine in to_tt_pass.py
Added tests for transpose in a separate test file "test_ops.py"

ayerofieiev-tt commented on tests/test_ops.py:
```diff
@@ -0,0 +1,60 @@
+        # The compilation is lazy, so we need to run forward once to trigger the compilation
+        m = torch.compile(m, backend=torch_ttnn.backend, options=option)
+        result_after = m.forward(input, dim0, dim1)
+        option._out_fx_graphs[0].print_tabular()
```
do we need this print?

mcw-zwakeelTT replied:
This printing is there for all existing tests, so I included it.

kevinwuTT replied:
Yeah all the tests have this from the start. I think it should be fine since it only shows up if we ask pytest to be verbose.

ayerofieiev-tt commented on tests/test_ops.py:
```diff
@@ -0,0 +1,60 @@
+
+        # Check the graph has be rewritten and contain ttnn ops
+        nodes = list(option._out_fx_graphs[0].nodes)
+        for node in nodes:
+            print(node.name)
```
do we need these?

mcw-zwakeelTT replied:
This we can remove, I forgot to remove at the last moment.

ayerofieiev-tt commented on torch_ttnn/passes/to_tt_pass.py:
```diff
@@ -368,6 +368,17 @@ def ReplaceMoreTtManually(gm: torch.fx.GraphModule) -> torch.fx.GraphModule:
                     new_node,
                     delete_user_cb=lambda node: node != new_node,
                 )
+            if node.target == torch.ops.aten.transpose.int:
```
what `int` stands for in `transpose.int`?

mcw-zwakeelTT replied:
Integer inputs. There could be other overloaded variants of transpose. This one is for integers.

ayerofieiev-tt replied:
how hard is it to handle non-integer transposes? why don't we just cover all variants here?

mcw-zwakeelTT replied:
It can be covered, have to check how many overloads are there and include them. I will capture this in the next commit.

mcw-zwakeelTT replied:
@ayerofieiev-tt I had a look at transpose overloads. There are two of them:
**torch.ops.aten.transpose.int** (uses integer indices to specify the dimensions to be transposed)
**torch.ops.aten.transpose.Dimname** (uses dimension names to specify the dimensions to be transposed)
Dimname is used for tensor dimension having names rather than integer indices. So int overload pretty much covers all input datatypes.

ayerofieiev-tt commented on torch_ttnn/passes/to_tt_pass.py:
```diff
@@ -368,6 +368,17 @@ def ReplaceMoreTtManually(gm: torch.fx.GraphModule) -> torch.fx.GraphModule:
                     new_node,
                     delete_user_cb=lambda node: node != new_node,
                 )
+            if node.target == torch.ops.aten.transpose.int:
```
Is there an in_place version of this operation?

mcw-zwakeelTT replied:
Not that I am aware of.

ayerofieiev-tt commented on torch_ttnn/passes/to_tt_pass.py:
```diff
@@ -368,6 +368,17 @@ def ReplaceMoreTtManually(gm: torch.fx.GraphModule) -> torch.fx.GraphModule:
+                new_node = g.call_function(ttnn.permute, args=(args[0], permutation))
+                node.replace_all_uses_with(
+                    new_node,
+                    delete_user_cb=lambda node: node != new_node,
```
what does this do?

mcw-zwakeelTT replied:
This replaces the users of existing torch node attaching to the users list of newly created ttnn node. Then the torch node can be deleted safely.

ayerofieiev-tt commented on tests/test_ops.py:
Whats the logic behind a name of this file? 
Lets call it `test_transpose.py` and keep it for a single op?

mcw-zwakeelTT replied:
test_more_ops.py contains tests for all existing ops. Thought the file was getting lengthy, so kept a separate file called it test_ops.py. We can merge it to have one single test python file, or if we want we can have separate files for individual ops. That way we can keep adding more variants of tests for a single op in future. But the downside could be too many files for tests.

kevinwuTT replied:
Yeah let's start separating ops into individual test files. There will be some repeated boiler-plate code, but it won't be too much. I will start refactoring the entire repo soon and reorganizing the tests will be part of it.

mcw-zwakeelTT replied:
Okay from now on, I will write separate tests for individual ops.

ayerofieiev-tt replied:
Please rename this file in this PR. Thanks

ayerofieiev-tt commented:
@mcw-zwakeelTT you can see that checks failed - thats because PR contains files which were not formatted and thats likely because you did not install pre-commit.

mcw-zwakeelTT commented:
> @mcw-zwakeelTT you can see that checks failed - thats because PR contains files which were not formatted and thats likely because you did not install pre-commit.

I have added a new commit covering this and renamed the test file.
